File: go/exercise/exercise_5.go
package exercises

import (
	"laba2/structs/huffman"
)

// HuffmanEncodeDecode выполняет кодирование и декодирование строки с помощью алгоритма Хаффмана
func HuffmanEncodeDecode(text string) (encoded string, decoded string, tree *huffman.HuffmanTree) {
	// Создаем дерево Хаффмана на основе входной строки
	tree = huffman.NewHuffmanTree(text)

	encoded = tree.Encode(text)

	decoded = tree.Decode(encoded)

	return encoded, decoded, tree
}

File: C++/exercise/exercise_5.cpp
#include "exercise_5.h"

using namespace std;

tuple<string, string, HuffmanTree*> huffmanEncodeDecode(const string& text) {
    // Создаем дерево Хаффмана на основе входной строки
    HuffmanTree* tree = new HuffmanTree(text);
    
    string encoded = tree->encode(text);
    
    string decoded = tree->decode(encoded);
    
    return make_tuple(encoded, decoded, tree);
}
